### PR TITLE
Add 'Blob.update_storage_class' API method.

### LINKED ISSUE #2991: Storage: add support for Per-Object Storage Class
We need to add support for Per-Object Storage Class (POSC).

Here some extra information to help:

- [POSC info from Cloud Storage documentation](https://cloud.google.com/storage/docs/per-object-storage-class)
- [POSC example with JSON API](https://cloud.google.com/storage/docs/changing-storage-classes)
- [Storage Class info from Cloud Storage documentation](https://cloud.google.com/storage/docs/storage-classes)

On top of that, google-cloud-dotnet has made an implementation which can be used as reference. 

- [Please view the Pull Request to understand the changes](https://github.com/GoogleCloudPlatform/google-cloud-dotnet/pull/766)

If you have any implementation questions, please direct them at jskeet

/cc @jskeet @danoscarmike

### PR DESCRIPTION
**Pull request by tseaver**
Closes #2991.

### CODE REVIEW

File: storage/google/cloud/storage/blob.py [hunk 1]
```diff
@@ -12,6 +12,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+# pylint: disable=too-many-lines
```
Comment (dhermes) [inline storage/google/cloud/storage/blob.py]: :sob:

File: storage/google/cloud/storage/blob.py [hunk 2]
```diff
@@ -73,6 +75,9 @@ class Blob(_PropertyMixin):
     _CHUNK_SIZE_MULTIPLE = 256 * 1024
     """Number (256 KB, in bytes) that must divide the chunk size."""
 
+    _STORAGE_CLASSES = (
+        'STANDARD', 'NEARLINE', 'MULTI_REGIONAL', 'REGIONAL', 'COLDLINE')
```
Comment (dhermes) [inline storage/google/cloud/storage/blob.py]: Don't we already have this [**somewhere**][1]?

[1]: https://github.com/GoogleCloudPlatform/google-cloud-python/blob/2c608f95d7d5db2cbfdce77770f523a489b513f4/storage/google/cloud/storage/bucket.py#L93-L94

File: storage/google/cloud/storage/blob.py [hunk 3]
```diff
@@ -852,6 +857,31 @@ def rewrite(self, source, token=None, client=None):
 
         return api_response['rewriteToken'], rewritten, size
 
+    def update_storage_class(self, new_class, client=None):
```
Comment (dhermes) [inline storage/google/cloud/storage/blob.py]: Shouldn't the constructor also get to set this?

File: storage/google/cloud/storage/blob.py [hunk 4]
```diff
@@ -852,6 +857,31 @@ def rewrite(self, source, token=None, client=None):
...
+        """Rewrite source blob into this one.
```
Comment (dhermes) [inline storage/google/cloud/storage/blob.py]: This isn't a very helpful one-liner

File: storage/google/cloud/storage/blob.py [hunk 5]
```diff
@@ -852,6 +857,31 @@ def rewrite(self, source, token=None, client=None):
...
+
+        :type new_class: str
+        :param new_class: new storage class for the object
+
+        :type client: :class:`~google.cloud.storage.client.Client` or
+                      ``NoneType``
```
Comment (dhermes) [inline storage/google/cloud/storage/blob.py]: The `or NoneType` isn't really needed since "Optional" conveys the same.

File: storage/google/cloud/storage/blob.py (hunk 1 above)
Comment (daspecster) [inline storage/google/cloud/storage/blob.py]: Should we make an issue/reminder to break this file up down the road?

I can make the issue if you think that's an ok plan.

File: storage/google/cloud/storage/blob.py [hunk 6]
```diff
@@ -852,6 +857,31 @@ def rewrite(self, source, token=None, client=None):
...
+        if new_class not in self._STORAGE_CLASSES:
+            raise ValueError("Invalid storage class: %s" % (new_class,))
+
+        client = self._require_client(client)
+        headers = _get_encryption_headers(self._encryption_key)
+        headers.update(_get_encryption_headers(
+            self._encryption_key, source=True))
+
+        api_response = client._connection.api_request(
+            method='POST', path=self.path + '/rewriteTo' + self.path,
```
Comment (dhermes) [inline storage/google/cloud/storage/blob.py]: Don't we need to also implement `rewrite` without the fancy storage class helper?

File: storage/google/cloud/storage/blob.py [hunk 7]
```diff
@@ -852,6 +857,31 @@ def rewrite(self, source, token=None, client=None):
...
+            data={'storageClass': new_class}, headers=headers,
```
Comment (dhermes) [inline storage/google/cloud/storage/blob.py]: No other data to send?

File: storage/unit_tests/test_blob.py [hunk 8]
```diff
@@ -1446,6 +1446,48 @@ def test_rewrite_same_name_no_key_new_key_w_token(self):
...
+        self.assertEqual(kw[0]['data'], SENT)
+
+        headers = {
+            key.title(): str(value) for key, value in kw[0]['headers'].items()}
+        self.assertNotIn('X-Goog-Copy-Source-Encryption-Algorithm', headers)
+        self.assertNotIn('X-Goog-Copy-Source-Encryption-Key', headers)
+        self.assertNotIn('X-Goog-Copy-Source-Encryption-Key-Sha256', headers)
+        self.assertNotIn('X-Goog-Encryption-Algorithm', headers)
+        self.assertNotIn('X-Goog-Encryption-Key', headers)
+        self.assertNotIn('X-Goog-Encryption-Key-Sha256', headers)
```
Comment (dhermes) [inline storage/unit_tests/test_blob.py]: 1. Why are these 6 keys hardcoded here?
1. All we know is what is **not** in the headers. What **is** in the headers?

File: storage/google/cloud/storage/blob.py (hunk 2 above)
Comment (tseaver) [inline storage/google/cloud/storage/blob.py]: The allowed set for blobs is smaller than for buckets.

File: storage/google/cloud/storage/blob.py (hunk 3 above)
Comment (tseaver) [inline storage/google/cloud/storage/blob.py]: Doesn't matter until we fix the "send metadata with upload" issues.

File: storage/google/cloud/storage/blob.py (hunk 6 above)
Comment (tseaver) [inline storage/google/cloud/storage/blob.py]: There is a separate `rewrite()` method already (added initially to support rotating encryption keys).

File: storage/google/cloud/storage/blob.py (hunk 7 above)
Comment (tseaver) [inline storage/google/cloud/storage/blob.py]: Nope. See the "REST APIs" section of https://cloud.google.com/storage/docs/changing-storage-classes

File: storage/unit_tests/test_blob.py (hunk 8 above)
Comment (tseaver) [inline storage/unit_tests/test_blob.py]: The blob has no encryption key set, and so we don't send any information to decrypt/re-encrypt it (check the similar tests for `rewrite()`).  The `headers` *might* be empty here, or they might have gotten injected with extra low-level stuff (like the GCCL stuff).

File: storage/google/cloud/storage/blob.py (hunk 4 above)
Comment (tseaver) [inline storage/google/cloud/storage/blob.py]: Yup, it is copy-pasta from the `rewrite` docstring.  Fixed in 087a7fb.

File: storage/google/cloud/storage/blob.py (hunk 5 above)
Comment (tseaver) [inline storage/google/cloud/storage/blob.py]: Fixed in 087a7fb.

File: storage/google/cloud/storage/blob.py (hunk 1 above)
Comment (tseaver) [inline storage/google/cloud/storage/blob.py]: Not specifically:  we hit an arbitrary, self-imposed limit.  Breaking the file up isn't necessarily a good thing.  You might cast the issue as "review those limits and the files where we disable them."

File: storage/google/cloud/storage/blob.py (hunk 3 above)
Comment (dhermes) [inline storage/google/cloud/storage/blob.py]: Ah. Blerg. We definitely need a tracking issue for that (kicking the can down the road is better than deleting it from our memory)

File: storage/google/cloud/storage/blob.py (hunk 6 above)
Comment (dhermes) [inline storage/google/cloud/storage/blob.py]: :+1:

File: storage/google/cloud/storage/blob.py (hunk 7 above)
Comment (dhermes) [inline storage/google/cloud/storage/blob.py]: Ack

File: storage/unit_tests/test_blob.py (hunk 8 above)
Comment (dhermes) [inline storage/unit_tests/test_blob.py]: Since we run the unit tests always, we can assert what is in the headers and if someone breaks this test by accident, then they can just fix it on the spot? (This is germane to a conversation @lukesneeringer and I were having earlier today.)

Comment (lukesneeringer) [inline storage/unit_tests/test_blob.py]: What is the test trying to establish? My personal feeling is that other, unrelated changes might alter headers, and asserting empty set seems like it is likely to end up breaking the test on unrelated changes.

If the only thing we need to prove is that these headers are absent, then what @tseaver has seems fine to me.

File: storage/google/cloud/storage/blob.py (hunk 3 above)
Comment (tseaver) [inline storage/google/cloud/storage/blob.py]: #754 and #1660 are open.

File: storage/unit_tests/test_blob.py (hunk 8 above)
Comment (tseaver) [inline storage/unit_tests/test_blob.py]: If it would make it clearer, I could add another test which showed updating the storage class for a blob which *does* have an encryption key:  in that case, those headers will be populated (with the same values for source and destination).

Comment (lukesneeringer) [inline storage/unit_tests/test_blob.py]: That sounds fine. (I am honestly fine with whatever in this case.)

Comment (tseaver) [inline storage/unit_tests/test_blob.py]: f1a03ea adds such a test, along with a comment in the no-encryption-key test.

File: storage/google/cloud/storage/blob.py (hunk 3 above)
Comment (dhermes) [inline storage/google/cloud/storage/blob.py]: I meant an issue to track adding the storage class to the constructor.

File: storage/google/cloud/storage/blob.py (hunk 2 above)
Comment (dhermes) [inline storage/google/cloud/storage/blob.py]: Are we sure that's true? Or has that list just been upgraded under our nose?

**Comment by dhermes** [comment]
Two remaining sticking points:

- A tracking issue for obj. storage class in the Blob constructor
- Making sure we don't repeat ourselves on the list of storage classes in blob and bucket modules

**Comment by tseaver** [comment]
@dhermes 

>  A tracking issue for obj. storage class in the Blob constructor

#3062.

>  Making sure we don't repeat ourselves on the list of storage classes in blob and bucket modules

The two sets aren't identical:  in particular, the DRA class is only documented for Buckets, although it is deprecated, while the 'STANDARD' class isn't documented for POSC at all (although it happens to work).

**Comment by lukesneeringer** [comment]
> The two sets aren't identical: in particular, the DRA class is only documented for Buckets, although it is deprecated, while the 'STANDARD' class isn't documented for POSC at all (although it happens to work).

Is one a strict superset of the other? If so, the file with the larger set could import the smaller one and `.union` on it. (You'd also have to change to sets rather than tuples, which would be fine.)

**Comment by tseaver** [comment]
> Is one a strict superset of the other? 

@lukesneeringer By inspection of the documented values at point-of-last-reading, yes:  however, the docs don't *specify* that:  they just describe the two sets which happen to overlap:

- The [`Bucket` resource spec](https://cloud.google.com/storage/docs/json_api/v1/buckets#storageClass) lists all the values currently laid out in our `Bucket._STORAGE_CLASSES` member.
- The [`Object` resource spec](https://cloud.google.com/storage/docs/json_api/v1/objects#storageClass) doesn't list any values at all (boo, hiss!)
- The [Storage Classes concept doc](https://cloud.google.com/storage/docs/storage-classes) lists only four (multi-regional, regional, nearline, and coldline), and [defines `STANDARD`](https://cloud.google.com/storage/docs/storage-classes#standard) as an alias based on the bucket's location type.  It [documents DRA](https://cloud.google.com/storage/docs/storage-classes#dra) as a deprecated version, pushing use of multi-regional or regional instead.
- The [Per-Object Storage Class concept docs](https://cloud.google.com/storage/docs/per-object-storage-class) *only* allow for the four types: I found by inspection that `STANDARD` is settable, and is the value which a new object shows in its resource if the bucket specified no storage class.

We could perhaps coalesce the two lists if we dropped support for setting the DRA type on buckets, or if we could get the storage back-end folks to document the possible values definitively for both buckets and objects.

**Comment by lukesneeringer** [comment]
Okay, got it. Based on that, I think I am going to say that we stick with @tseaver's approach as implemented for now (while acknowledging that it is a close call).

**Comment by dhermes** [comment]
@lukesneeringer @tseaver We should

- include a comment in code explaining why the duplication
- give feedback to the backend folks that we had to duplicate / the choices don't seem clear from the docs (/cc @jonparrott)

**Comment by tseaver** [comment]
@dhermes c34cd86 adds docstrings to both class attributes, including links to the relevant back-end documentation.

**Comment by tseaver** [comment]
@dhermes, @lukesneeringer Any further issues?

**Comment by lukesneeringer** [comment]
This looks good. For some reason I cannot find the button to change my review status to "approved", so consider this comment to be tacit approval. :-p
